### ISSUE TITLE
Two-way communication with the Swift binary

### ISSUE THREAD

**Issue by sindresorhus** [wulkano/aperture#47]
We'll need two-way communication to be able to implement pause/resume recording (#28) and to fix #40. We need to use some kind of socket/IPC communication (https://github.com/wulkano/aperture/issues/8#issuecomment-277975552).

**Comment by sindresorhus** [comment]
I just thought of a possible solution. We could use [`NSDistributedNotificationCenter`](https://developer.apple.com/documentation/foundation/distributednotificationcenter) to communicate between processes.

So on the Swift side we would send a notification when the Swift binary actually started recording:

```swift
DistributedNotificationCenter.default().postNotificationName(NSNotification.Name("aperture.recording"), object: nil, userInfo: ["file": url], deliverImmediately: true)
```

And on the Electron-side we would listen for that event:

```js
electron.systemPreferences.subscribeNotification('aperture.recording', (event, userInfo) => {
	console.log('event', event, userInfo);
});
```

Which would output the following from Electron:

```
event aperture.recording { file: '…' }
```

This would work both ways. The downside is that we would need to write it specifically for Electron, so it would need to be a different mode/option in Aperture.

---

@albinekb @LarsJK Thoughts?

**Comment by LarsJK** [comment]
Yeah seems like a good solution to me, although I'm not too familiar with DistributedNotificationCenter or Electron..
Maybe you could use libnotify like that last article mentions as a C alternative to DistibutedNotificationCenter? Seems like there are a few node packages for libnotify..

I don't know how things like node-gyp or node-ffi works but would you be able to call functions if you made a system-module in swift?

**Comment by sindresorhus** [comment]
@LarsJK Ah yeah, that's an interesting idea. I couldn't find a Node.js module for this, but shouldn't be too hard to create a simple one that interfaces `libnotify` using FFI.

**Comment by LarsJK** [comment]
Maybe [XPC](https://www.objc.io/issues/14-mac/xpc/) could be an alternative as well..

Haven't actually used any of these myself, so I don't know which one would be best..

**Comment by albinekb** [comment]
Perhaps that lib could use the electron implementation if it's available? 🤔  @sindresorhus But yes, that sounds like the best solution so far.
